Validate company application page

FranciscoCardoso913 opened the pull request:
### Important:

This is a backend issue! See the original issue and the changes made in the backend here: https://github.com/NIAEFEUP/nijobs-be/pull/306


### Discussed solution:
    
    -When created, applications should have a state UNVERIFIED by default instead of the current PENDING state;

    -After creating an application an email with a link should be sent to the company so they can validate their application, the link should expire in 10 minutes;

    -A company should not be able to create an application with an email associated with an UNVERIFIED application whose link is yet to expire;

    -If the link of an application expires and the company tries to create another application the previous application of that company should be deleted;

    -The link should send the user to a page where the application is validated, when validating the application its state should change to pending and an email should be sent to the company.

    

### Front-End changes:

    -You can now search for unverified applications;
![Screenshot from 2023-02-11 20-18-26](https://user-images.githubusercontent.com/94190384/218279473-b126057a-5c8d-44ba-be92-08dbeb9016d2.png)

    -The message which appeared after creating an application has now changed to indicate the company they have to validate their application;
![Screenshot from 2023-02-11 20-14-37](https://user-images.githubusercontent.com/94190384/218279352-f9debec6-27b5-4e57-8526-188bd1e70150.png)

    -New error message when trying to create a new application with an email associated with an UNVERIFIED application whose link is yet to expire;
![Screenshot from 2023-02-11 20-34-35](https://user-images.githubusercontent.com/94190384/218280055-eb20c47f-c13c-4b1c-8b3e-7b65ce4de8c9.png)


    -Created a new page for validating the company, the page displays various messages indicating if the validation was successful or if an error has occurred;

### States of the new page:

    -Page when loading:

![Screenshot from 2023-02-11 20-22-53](https://user-images.githubusercontent.com/94190384/218279640-033a6ffc-159c-4fa0-ac64-f2fd58e8c003.png)


    -Page when everything goes well:

![Screenshot from 2023-02-11 20-25-06](https://user-images.githubusercontent.com/94190384/218279702-3071861d-40db-4663-ba18-6a0f35208276.png)


     -Page when the application was already validated:

![Screenshot from 2023-02-11 20-30-44](https://user-images.githubusercontent.com/94190384/218279906-9c55b387-6263-4643-9a41-e7f36ca23a18.png)


    -Page when the application does not exist(the token in the link is incorret):

![Screenshot from 2023-02-11 20-27-41](https://user-images.githubusercontent.com/94190384/218279805-1a46ce9e-2e51-4bbe-b138-83794b636248.png)


    -Page when the link has expired:

![Screenshot from 2023-02-11 20-26-34](https://user-images.githubusercontent.com/94190384/218279758-dd2ed3f2-bfb3-484c-8fef-76981d56ee6c.png)


    -Page when an unexpected error occurs:

![Screenshot from 2023-02-11 20-28-51](https://user-images.githubusercontent.com/94190384/218279852-9b8f5eae-28ca-4333-aa7a-58d8a088d747.png)



### What is yet to be done:

     -Tests;

     -Google analytics;
    

### Doubts:

    -The error messages used have a style made by us, however, there already exists a template style to display errors ( a component called ErrorComponent), should we use that or keep it as it is?

    -Should we change any of the designs of the validation page?

    -Should we change any of the texts?	

    -Is everything in the correct place or is there anything that is more adequate somewhere else?

FranciscoCardoso913 committed:
Added validation of company application route and unverified state

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

FranciscoCardoso913 committed:
Implemented validation page design

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

FranciscoCardoso913 committed:
Added appropriate error messages for application

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

DoStini commented on src/components/Apply/Company/CompanyApplicationUtils.js:
```diff
@@ -19,6 +17,34 @@ export const generateValidationRule = validationRulesGenerator(CompanyApplicatio
+        title: "Error! Application does not exist!",
+        text: "An error has occured while validating your application! The application you are trying to validate does not exist,",
+    },
+    "expired-token": {
+        title: "Error! Link has expired!",
+        text: "An error has occured while validating your application! The link sent to you has expired, \
+        you now need to create a new application,",
+    },
```
Now that I'm thinking about this, an expired token should be considered as a NOT FOUND application. This is more backend wise

DoStini replied:
Don't use a title, put these in the HumanReadableErrors

FranciscoCardoso913 replied:
Our ideia was to make it easier for the company to identify what went wrong, that way they can see if the token has expired or if the token is just wrong, otherwise it will seem as it is the token is wrong when in reality it has only expired.
Okok we will change that

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+import { validateApplication } from "../services/companyApplicationService";
+import { getValidationError } from "../components/Apply/Company/CompanyApplicationUtils.js";
+
+const useStyles = (isMobile) => makeStyles((theme) => ({
+    content: {
+        padding: isMobile ? theme.spacing(2, 2) : theme.spacing(3, 3),
+        display: "flex",
+        flexDirection: "column",
+        alignItems: "center",
+        justifyContent: "center",
+    },
+    title: {
+        color: "rgb(80,80,80)",
+        fontSize: "23px",
+    },
+    text: {
+        color: "rgb(100,100,100)",
+        fontSize: "15px",
+        maxWidth: isMobile ? "100%" : "80%",
+        textAlign: "center",
+
+    },
+    button: {
+        background: "rgb(250,70,70)",
+        color: "white",
+        marginTop: "3vh",
+
+    },
+}));
```
This looks like way too much CSS, we will have to analyse this to see what can be replace with Material UI defaults

Naapperas replied:
Also, `makeStyles` returns a function that can optionally take arguments, which in this case would be `isMobile`, so there is no need to apply currying here: that argument is available to use if instead of an object you use a function as the style, for instance:

Instead of
```js
text: {
    ...
    maxWidth: isMobile ? "100%" : "80%",
}
```
you could have
```js
text: ({isMobile}) => ({
    ...
    maxWidth: isMobile ? "100%" : "80%",
})
```
which makes it so only the styles that may need it use it.

I'm not entirely sure about the syntax but I believe this is a thing. Correct me if I'm wrong.

This is not so much a problem as it is just advice/letting you know.

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+        justifyContent: "center",
+    },
+    title: {
+        color: "rgb(80,80,80)",
```
Use theme colors

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+    },
+    title: {
+        color: "rgb(80,80,80)",
+        fontSize: "23px",
```
Hard coded font size, simply use `<Typography>` with `variant` as `h1,h2,...`

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+        fontSize: "23px",
+    },
+    text: {
+        color: "rgb(100,100,100)",
```
hardcoded color

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+    },
+    text: {
+        color: "rgb(100,100,100)",
+        fontSize: "15px",
```
hardcoded font size -> non responsive

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+
+    },
+    button: {
+        background: "rgb(250,70,70)",
+        color: "white",
```
same as colors

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+
+
+    useEffect(() => {
+        async function validate() {
+            try {
+                setLoading(false);
+                setSuccess(true);
+                await validateApplication(token);
+            } catch (error_) {
+                setError(error_[0].msg);
+                setLoading(false);
+                setSuccess(false);
+            }
+
+        }
```
Why are you creating a method and calling it immediately after?

FranciscoCardoso913 replied:
UseEffect can't be async, because of that if you want to have an await inside of an useEffect you have to use an auxiliar async function , .then() would also work as an alternative

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+                setLoading(false);
+                setSuccess(true);
+                await validateApplication(token);
+            } catch (error_) {
```
```suggestion
            } catch (err) {
```

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+                    <h2 className={classes.title}>Your application has been validated successfully! </h2>
+                    <span className={classes.text}>
+                        You should receive a confirmation email shortly. If not, please contact us:
+                         <a href="mailto:nijobs@aefeup.pt"> nijobs@aefeup.pt</a>
```
I Think there's a constant for this somewhere, please check it

DoStini replied:
Use `<Link>` component with href

DoStini commented on src/pages/ValidationPage.js:
```diff
@@ -0,0 +1,113 @@
+                        You should receive a confirmation email shortly. If not, please contact us:
+                         <a href="mailto:nijobs@aefeup.pt"> nijobs@aefeup.pt</a>
+                    </span>
+                    <Button className={classes.button} variant="contained" onClick={() => setButtonPressed(true) }>
+                        Click here to go to Home page
+                    </Button>
```
Use a `<Link>` component instead of state and redirect. Use with RouterLink component. Check other examples

DoStini commented on src/components/Review/Applications/ApplicationsReviewTableSchema.js:
```diff
@@ -6,7 +6,9 @@ export const columns = {
 };
 
 export const ApplicationStateLabel = Object.freeze({
+    UNVERIFIED: "Unverified",
```
Do we want Unverified aaplications to show uo in the admin panel? I think it shouldn't

- Still not a valid application
- If we are victims of spam, they will show up

DoStini replied:
cc @BrunoRosendo @Naapperas For more opinions

BrunoRosendo replied:
What's the flow here? We can only accept them once they've confirmed their emails? Imo that's too much friction for a website where it's already hard to bring companies into. They need to register, confirm email, wait for the application, and finish registration before starting to create offers.

What if, instead of that, we let the companies use the website without confirming email (just like today), and use the `isVerified` prop like a Twitter verification? The company could have a warning (snack bar) after logging in and the students (and admins) would be able to know if the company is fully trust worthy.

FranciscoCardoso913 replied:
If i am not mistaken, in a reunion it was discussed that Unverified should be a state like PENDING APROVED and REJECTED , so you can filter your application to show unverified application or others. About the spam, as it stands only one application with the same email can exist , however that way we are still vonorable to be spamed with applications with different emails, one solution to that would be deleting all application that are unverified after 10 minutes (when you can longer validate them) , but of course that way having an unverified state would no longer make sense.

As it stands when creating an application your application will be Unverified and cannot be reviewed, you will than receive an email to verifie it, if you do so then your application become PENDING and is ready to be reviewed. 

The idea of being like twitter is an interesting ideia, however it may be better to discuss it with the rest of the team as it is a concept completely different from the original ideia( it may be better to discussed it on the next reunion). In that case isVerified should probably be on the company instead of the application and the way companies verifie their companies should be diferent and only done after they are aproved. The email with a link that expires in 10 minutes which is sent when the application is created would no longer make a lot of sense in that case, maybe we could have a button in the company profile where you could validate your company, that way we could still use the email with the link to validate it

BrunoRosendo replied:
Yes, I think we should discuss this again. I know it wasn't the original idea but I was re-thinking about it myself

dsantosferreira replied:
> What's the flow here? We can only accept them once they've confirmed their emails? Imo that's too much friction for a website where it's already hard to bring companies into. They need to register, confirm email, wait for the application, and finish registration before starting to create offers.
> 
> What if, instead of that, we let the companies use the website without confirming email (just like today), and use the `isVerified` prop like a Twitter verification? The company could have a warning (snack bar) after logging in and the students (and admins) would be able to know if the company is fully trust worthy.

I really like the Twitter verification idea but there is something that is bothering me. I'm sorry if what I'm about to say is completely wrong, but wouldn't the spam issue persist in the offers page, as each account can potentially create up to five concurrent offers? Wouldn't the offer page or even our database become way too bloated?
An idea that i had was to develop a scheduled function that would run every hour or so (not sure of the appropriate time interval) that would erase all expired applications.

Naapperas replied:
I like the Twitter thing, in addition to said pop-up/snackbar, perhaps we could/should restrict some actions made by unverified companies, but this may cause more harm than good. It should be discussed in the weekly meting but I personally am leaning towards what @BrunoRosendo suggested.

DoStini commented on src/AppRouter.js:
```diff
@@ -135,6 +136,15 @@ const AppRouter = () => (
+            <Route exact path="/apply/validate/:token/confirm" key="/apply/validate/:token/confirm">
+                <PageLayout
+                    key="/apply/validate/:token/confirm"
+                    pageTitle="Validate"
```
```suggestion
                    pageTitle="Validate Company Application"
```

FranciscoCardoso913 commented:
## Important
The workflow regarding this issue has been changed
You can check it here: https://github.com/NIAEFEUP/nijobs-be/pull/306#issuecomment-1447121249

FranciscoCardoso913 committed:
Changed hyperlinks and redirect to Link component and changed text to Typography

Naapperas commented:
@FranciscoCardoso913 @dsantosferreira any progress on this?

FranciscoCardoso913 committed:
Solved redirected to home page issue and refactored validation page code

FranciscoCardoso913 committed:
Added a waring to MyOffers page and to the CreateOffer page when company is yet to be approved

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

FranciscoCardoso913 committed:
Added a waring to MyOffers page and to the CreateOffer page when company is yet to be approved

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

FranciscoCardoso913 committed:
Merge remote-tracking branch 'origin/validate-company-application-Page' into validate-company-application-Page

dsantosferreira commented:
We have our first version of the warnings shown to users whose accounts aren't yet verified in the "Create an offer" and "My offers" pages. Feel free to share your opinions!

![warning1](https://github.com/NIAEFEUP/nijobs-fe/assets/96185733/a62d5f88-3d78-4961-8091-9cccf639e7e2)
![warning2](https://github.com/NIAEFEUP/nijobs-fe/assets/96185733/39710321-66d9-402f-8d15-8d5ab745d923)

BrunoRosendo commented:
> We have our first version of the warnings shown to users whose accounts aren't yet verified in the "Create an offer" and "My offers" pages. Feel free to share your opinions!
> 
> ![warning1](https://user-images.githubusercontent.com/96185733/247513980-a62d5f88-3d78-4961-8091-9cccf639e7e2.jpeg) ![warning2](https://user-images.githubusercontent.com/96185733/247513985-39710321-66d9-402f-8d15-8d5ab745d923.jpeg)

I like how it turned out! I just have some doubts about the icon. Is it also from material UI? The style seems a bit off from the rest of your icons, is it imported from `@material-ui/icons` or somewhere else?

Assuming it is the right one, I guess it's the one on the right. Do you mind trying out the first one and see how it looks like?
![image](https://github.com/NIAEFEUP/nijobs-fe/assets/55768934/4fdbaccc-5aef-4a05-997b-1cefd48e59d3)

dsantosferreira commented:
> 

The icon on the images I sent is the default one used by the alert component from material UI when the severity attribute is set to "warning". It is probably the "WarningAmber" icon, as you referred.
Here is the alert with the filled version of the icon. I prefer the outlined one but tell me what you think!

![image](https://github.com/NIAEFEUP/nijobs-fe/assets/96185733/21467778-de8b-4b66-8803-6375d13a164c)

Naapperas commented:
> > 
> 
> The icon on the images I sent is the default one used by the alert component from material UI when the severity attribute is set to "warning". It is probably the "WarningAmber" icon, as you referred. Here is the alert with the filled version of the icon. I prefer the outlined one but tell me what you think!
> 
> ![image](https://user-images.githubusercontent.com/96185733/248418198-21467778-de8b-4b66-8803-6375d13a164c.png)

Honestly, I prefer the filed version of the icon.

BrunoRosendo commented:
> > 
> 
> The icon on the images I sent is the default one used by the alert component from material UI when the severity attribute is set to "warning". It is probably the "WarningAmber" icon, as you referred. Here is the alert with the filled version of the icon. I prefer the outlined one but tell me what you think!
> 
> ![image](https://user-images.githubusercontent.com/96185733/248418198-21467778-de8b-4b66-8803-6375d13a164c.png)

I do think the filled version looks more like the remaining icons we use (from what I recall, but feel free to check it out)

dsantosferreira committed:
Conditionally render the alert for non yet approved companies and change warning icon to its filled version

Co-authored-by: Francisco Cardoso <FranciscoCardoso913@users.noreply.github.com>

FranciscoCardoso913 committed:
Rewriting the website messages to match the current flow

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

FranciscoCardoso913 committed:
Rewriting the website messages to match the current flow

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

Naapperas reviewed (changes requested):
Remove the `console.log` statements

FranciscoCardoso913 committed:
Fixing failing tests

dsantosferreira committed:
Created test to check if approve and reject buttons are not rendered in unverified applications

FranciscoCardoso913 committed:
Tests for validation page

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

FranciscoCardoso913 committed:
Tests for validation page

Co-authored-by: Daniel Ferreira <dsantosferreira@users.noreply.github.com>

FranciscoCardoso913 committed:
Tests for the alerts components in the company offer management page and the create offer form

FranciscoCardoso913 committed:
Making tests for the validate application and the fetch company application services

Making tests for the validate application and the fetch company application services

Naapperas commented on src/components/Apply/Company/CompanyApplicationUtils.js:
```diff
@@ -19,6 +17,38 @@ export const generateValidationRule = validationRulesGenerator(CompanyApplicatio
+            " You will need to create a new application.",
+    },
+    "application-already-validated": {
+        title: "Application is already validated!",
+        text: "This application is already validated. ",
```
```suggestion
        title: "Application was already validated!",
        text: "This application was already validated. ",
```

FranciscoCardoso913 committed:
Add extra error message to validation page, fixed: can enter create offer page with admin account

FranciscoCardoso913 committed:
Merge branch 'develop' into validate-company-application-Page
